enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java

```diff
@@ -0,0 +1,169 @@
+        {
+            try
+            {
+                relationshipMap.put( propertyEntry.entity().getId(), propertyEntry.entity().getProperties( relationshipProperties ) );
+            }
+            catch ( NotFoundException e )
+            {
+                //This means that the relationship was deleted.
+            }
+        } );
+        data.assignedRelationshipProperties().forEach(
+                propertyEntry -> relationshipMap.put( propertyEntry.entity().getId(), propertyEntry.entity().getProperties( relationshipProperties ) ) );
+        return new Map[]{nodeMap, relationshipMap};
+    }
+
+    @Override
+    public void afterCommit( TransactionData data, Object state )
+    {
+        //update node indices
+        try
+        {
+            for ( WritableFulltext nodeIndex : fulltextProvider.writableNodeIndices() )
+            {
+                Map<Long,Map<String,Object>> nodeMap = ((Map<Long,Map<String,Object>>[]) state)[0];
+                removePropertyData( data.removedNodeProperties(), nodeMap, nodeIndex );
+                updatePropertyData( nodeMap, nodeIndex );
+                refreshIndex( nodeIndex );
+            }
+            //update relationship indices
+            for ( WritableFulltext relationshipIndex : fulltextProvider.writableRelationshipIndices() )
+            {
+                Map<Long,Map<String,Object>> relationshipMap = ((Map<Long,Map<String,Object>>[]) state)[1];
+                removePropertyData( data.removedRelationshipProperties(), relationshipMap, relationshipIndex );
+                updatePropertyData( relationshipMap, relationshipIndex );
+                refreshIndex( relationshipIndex );
+            }
+        }
+        catch ( IOException e )
+        {
+            log.error( "Unable to update fulltext index", e );
+        }
+    }
+
+    private <E extends Entity> void updatePropertyData( Map<Long,Map<String,Object>> state, WritableFulltext index ) throws IOException
+    {
+        for ( Map.Entry<Long,Map<String,Object>> stateEntry : state.entrySet() )
+        {
+            Set<String> indexedProperties = index.properties();
+            if ( !Collections.disjoint( indexedProperties, stateEntry.getValue().keySet() ) )
+            {
+                long entityId = stateEntry.getKey();
+                Map<String,Object> allProperties =
+                        stateEntry.getValue().entrySet().stream().filter( entry -> indexedProperties.contains( entry.getKey() ) ).collect(
+                                Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) );
+                if ( !allProperties.isEmpty() )
+                {
+                    Document document = LuceneFulltextDocumentStructure.documentRepresentingProperties( entityId, allProperties );
+                    index.getIndexWriter().updateDocument( LuceneFulltextDocumentStructure.newTermForChangeOrRemove( entityId ), document );
```

Comment:
we can reuse same writer for all of the updates

Reply:
We also noticed that. I think we'll do it in a different PR because we already have a few changes lined up for this part of the code.